gpcontrib/arenadata_toolkit/arenadata_toolkit--1.4--1.5.sql

```diff
@@ -0,0 +1,125 @@
+/* gpcontrib/arenadata_toolkit/arenadata_toolkit--1.4--1.5.sql */
+
+/*
+ This is part of arenadata_toolkit API for ADB Bundle.
+ This function collects information in db_files_current and db_files_history tables.
+ */
+CREATE OR REPLACE FUNCTION arenadata_toolkit.adb_collect_table_stats()
+RETURNS VOID
+AS $$
+BEGIN
+	IF NOT EXISTS (
+		SELECT
+		FROM pg_partition pp
+		JOIN pg_class cl1 ON pp.parrelid = cl1.oid
+		JOIN pg_partition_rule pr ON pr.paroid = pp.oid
+		JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
+		JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
+		JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
+		WHERE
+			CASE
+				WHEN pp.parclass[0] IN (SELECT oid FROM pg_opclass WHERE opcname = 'timestamp_ops')

```

Comment:
why not
```sql
join pg_opclass on oid = parclass[0]
```
and
```sql
case when opcname = 'timestamp_ops' then
```

Reply:
I've created 1000 partitioned tables in `public` schema and 1000 partitioned tables in `arenadata_toolkit` schema, and launched these 2 options of the query.
The execution times in my experiment significantly differ.

`Execution time: 16.642 ms` (option with subselect in `case`) VS `Execution time: 145.421 ms` (option with pg_opclass join)

Details are below:
<details>
<summary>explain analyze logs</summary>

```
postgres=# explain analyze SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                WHERE
                        CASE
                            WHEN
                                                        pp.parclass[0] IN (SELECT oid FROM pg_opclass WHERE opcname = 'timestamp_ops')
                            THEN now() BETWEEN 
                                        CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                        AND
                                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                            ELSE FALSE
                        END
                                                AND
                                                cl1.relname = 'db_files_history'
                                                AND
                        n1.nspname = 'arenadata_toolkit'
                                                AND
                        n2.nspname = 'arenadata_toolkit'
                        AND
                        pp.paristemplate = false
                        AND
                        pp.parkind = 'r'  --range
                        AND
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                                 QUERY PLA
N                                                                                                                                                                                                                                  
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=227.63..391.26 rows=4 width=0) (actual time=12.624..16.405 rows=1 loops=1)
   Hash Cond: (cl2.relnamespace = n2.oid)
   Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
   ->  Hash Join  (cost=226.53..390.11 rows=4 width=4) (actual time=12.382..16.023 rows=1 loops=1)
         Hash Cond: (cl2.oid = pr.parchildrelid)
         Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
         ->  Seq Scan on pg_class cl2  (cost=0.00..142.94 rows=5494 width=8) (actual time=0.008..1.968 rows=3516 loops=1)
         ->  Hash  (cost=226.49..226.49 rows=2 width=4) (actual time=12.219..12.219 rows=1 loops=1)
               Buckets: 131072  Batches: 1  Memory Usage: 1kB
               ->  Hash Join  (cost=69.66..226.49 rows=4 width=4) (actual time=7.681..12.211 rows=1 loops=1)
                     Hash Cond: (cl1.relnamespace = n1.oid)
                     Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
                     ->  Hash Join  (cost=68.56..225.34 rows=4 width=8) (actual time=7.576..11.758 rows=1 loops=1)
                           Hash Cond: (pp.oid = pr.paroid)
                           Join Filter: CASE WHEN (hashed SubPlan 1) THEN (('2024-04-11 13:27:10.386977+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::time
stamp without time zone) AND ('2024-04-11 13:27:10.386977+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)) ELSE false END
                           Rows Removed by Join Filter: 2
                           Extra Text: Hash chain length 2.0 avg, 4 max, using 1002 of 32768 buckets.
                           ->  Hash Join  (cost=7.01..163.72 rows=4 width=40) (actual time=0.543..4.382 rows=2 loops=1)
                                 Hash Cond: (cl1.oid = pp.parrelid)
                                 Extra Text: Hash chain length 1.0 avg, 2 max, using 1008 of 65536 buckets.
                                 ->  Seq Scan on pg_class cl1  (cost=0.00..156.68 rows=1 width=8) (actual time=0.009..3.494 rows=2 loops=1)
                                       Filter: (relname = 'db_files_history'::name)
                                       Rows Removed by Filter: 3514
                                 ->  Hash  (cost=7.00..7.00 rows=1 width=40) (actual time=0.483..0.483 rows=1015 loops=1)
                                       Buckets: 65536  Batches: 1  Memory Usage: 64kB
                                       ->  Seq Scan on pg_partition pp  (cost=0.00..7.00 rows=1 width=40) (actual time=0.027..0.349 rows=1015 loops=1)
                                             Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
                           ->  Hash  (cost=58.00..58.00 rows=1 width=72) (actual time=1.086..1.086 rows=2029 loops=1)
                                 Buckets: 32768  Batches: 1  Memory Usage: 721kB
                                 ->  Seq Scan on pg_partition_rule pr  (cost=0.00..58.00 rows=1 width=72) (actual time=0.007..0.718 rows=2029 loops=1)
                                       Filter: parrangestartincl
                                       Rows Removed by Filter: 1
                           SubPlan 1
                             ->  Materialize  (cost=0.00..3.54 rows=4 width=4) (actual time=0.024..0.025 rows=3 loops=1)
                                   ->  Seq Scan on pg_opclass  (cost=0.00..3.53 rows=4 width=4) (actual time=0.008..0.020 rows=3 loops=1)
                                         Filter: (opcname = 'timestamp_ops'::name)
                                         Rows Removed by Filter: 199
                     ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.005..0.005 rows=1 loops=1)
                           Buckets: 131072  Batches: 1  Memory Usage: 1kB
                           ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4) (actual time=0.003..0.004 rows=1 loops=1)
                                 Filter: (nspname = 'arenadata_toolkit'::name)
                                 Rows Removed by Filter: 7
   ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.016..0.016 rows=1 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1kB
         ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4) (actual time=0.011..0.012 rows=1 loops=1)
               Filter: (nspname = 'arenadata_toolkit'::name)
               Rows Removed by Filter: 7
 Planning time: 4.845 ms
   (slice0)    Executor memory: 16713K bytes.  Work_mem: 721K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 16.642 ms
(52 rows)

postgres=# 
postgres=# explain analyze SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                                JOIN pg_opclass po on pp.parclass[0] = po.oid
                WHERE
                        CASE
                            WHEN
                                                        po.opcname = 'timestamp_ops'
                            THEN now() BETWEEN 
                                        CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                        AND
                                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                            ELSE FALSE
                        END
                                                AND
                                                cl1.relname = 'db_files_history'
                                                AND
                        n1.nspname = 'arenadata_toolkit'
                                                AND
                        n2.nspname = 'arenadata_toolkit'
                        AND
                        pp.paristemplate = false
                        AND
                        pp.parkind = 'r'  --range
                        AND
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                                 QUERY PLA
N                                                                                                                                                                                                                                  
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=227.93..391.92 rows=8 width=0) (actual time=27.624..144.992 rows=1 loops=1)
   Hash Cond: (pp.parrelid = cl1.oid)
   Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
   ->  Hash Join  (cost=70.07..233.94 rows=8 width=4) (actual time=21.849..138.595 rows=1002 loops=1)
         Hash Cond: (pr.paroid = pp.oid)
         Join Filter: CASE WHEN (po.opcname = 'timestamp_ops'::name) THEN (('2024-04-11 13:27:32.137083+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::time
stamp without time zone) AND ('2024-04-11 13:27:32.137083+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)) ELSE false END
         Rows Removed by Join Filter: 1025
         Extra Text: Hash chain length 1.0 avg, 2 max, using 1002 of 32768 buckets.
         ->  Hash Join  (cost=59.11..222.86 rows=4 width=72) (actual time=5.912..10.990 rows=2027 loops=1)
               Hash Cond: (cl2.relnamespace = n2.oid)
               Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
               ->  Hash Join  (cost=58.01..221.68 rows=13 width=76) (actual time=5.778..10.001 rows=2029 loops=1)
                     Hash Cond: (cl2.oid = pr.parchildrelid)
                     Extra Text: Hash chain length 1.0 avg, 3 max, using 1961 of 32768 buckets.
                     ->  Seq Scan on pg_class cl2  (cost=0.00..142.94 rows=5494 width=8) (actual time=0.016..2.098 rows=3516 loops=1)
                     ->  Hash  (cost=58.00..58.00 rows=1 width=72) (actual time=5.609..5.609 rows=2029 loops=1)
                           Buckets: 32768  Batches: 1  Memory Usage: 721kB
                           ->  Seq Scan on pg_partition_rule pr  (cost=0.00..58.00 rows=1 width=72) (actual time=0.014..1.777 rows=2029 loops=1)
                                 Filter: parrangestartincl
                                 Rows Removed by Filter: 1
               ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.022..0.022 rows=1 loops=1)
                     Buckets: 131072  Batches: 1  Memory Usage: 1kB
                     ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4) (actual time=0.019..0.019 rows=1 loops=1)
                           Filter: (nspname = 'arenadata_toolkit'::name)
                           Rows Removed by Filter: 7
         ->  Hash  (cost=10.87..10.87 rows=3 width=72) (actual time=1.459..1.459 rows=1015 loops=1)
               Buckets: 32768  Batches: 1  Memory Usage: 96kB
               ->  Hash Join  (cost=7.01..10.87 rows=8 width=72) (actual time=0.706..1.161 rows=1015 loops=1)
                     Hash Cond: (po.oid = pp.parclass[0])
                     Extra Text: Hash chain length 507.5 avg, 1002 max, using 2 of 65536 buckets.
                     ->  Seq Scan on pg_opclass po  (cost=0.00..3.02 rows=202 width=68) (actual time=0.015..0.072 rows=202 loops=1)
                     ->  Hash  (cost=7.00..7.00 rows=1 width=40) (actual time=0.624..0.624 rows=1015 loops=1)
                           Buckets: 65536  Batches: 1  Memory Usage: 64kB
                           ->  Seq Scan on pg_partition pp  (cost=0.00..7.00 rows=1 width=40) (actual time=0.017..0.368 rows=1015 loops=1)
                                 Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
   ->  Hash  (cost=157.81..157.81 rows=2 width=4) (actual time=5.570..5.570 rows=1 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1kB
         ->  Hash Join  (cost=1.10..157.81 rows=4 width=4) (actual time=0.535..5.541 rows=1 loops=1)
               Hash Cond: (cl1.relnamespace = n1.oid)
               Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
               ->  Seq Scan on pg_class cl1  (cost=0.00..156.68 rows=1 width=8) (actual time=0.103..4.324 rows=2 loops=1)
                     Filter: (relname = 'db_files_history'::name)
                     Rows Removed by Filter: 3514
               ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.076..0.076 rows=1 loops=1)
                     Buckets: 131072  Batches: 1  Memory Usage: 1kB
                     ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4) (actual time=0.055..0.057 rows=1 loops=1)
                           Filter: (nspname = 'arenadata_toolkit'::name)
                           Rows Removed by Filter: 7
 Planning time: 9.686 ms
   (slice0)    Executor memory: 14449K bytes.  Work_mem: 721K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 145.421 ms
(53 rows)

postgres=# 

```

</details>

Do you need any other experiment?

Reply:
I've created 1000 partitioned tables in `public` schema and 1000 partitioned tables in `arenadata_toolkit` schema, and launched these 2 options of the query.
The execution times in my experiment significantly differ.

**Iteration 1:**
`Execution time: 16.642 ms` (option with subselect in `case`) VS `Execution time: 145.421 ms` (option with pg_opclass join)

**Iteration 2:**
`Execution time: 9.024 ms` (option with subselect in `case`) VS `Execution time: 118.728 ms` (option with pg_opclass join)

Details are below:
<details>
<summary>explain analyze logs</summary>

```

postgres=# explain analyze SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                WHERE
                        CASE
                            WHEN
                                                        pp.parclass[0] IN (SELECT oid FROM pg_opclass WHERE opcname = 'timestamp_ops')
                            THEN now() BETWEEN 
                                        CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                        AND
                                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                            ELSE FALSE
                        END
                                                AND
                                                cl1.relname = 'db_files_history'
                                                AND
                        n1.nspname = 'arenadata_toolkit'
                                                AND
                        n2.nspname = 'arenadata_toolkit'
                        AND
                        pp.paristemplate = false
                        AND
                        pp.parkind = 'r'  --range
                        AND
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                                 QUERY PLA
N                                                                                                                                                                                                                                  
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=227.63..391.26 rows=4 width=0) (actual time=12.624..16.405 rows=1 loops=1)
   Hash Cond: (cl2.relnamespace = n2.oid)
   Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
   ->  Hash Join  (cost=226.53..390.11 rows=4 width=4) (actual time=12.382..16.023 rows=1 loops=1)
         Hash Cond: (cl2.oid = pr.parchildrelid)
         Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
         ->  Seq Scan on pg_class cl2  (cost=0.00..142.94 rows=5494 width=8) (actual time=0.008..1.968 rows=3516 loops=1)
         ->  Hash  (cost=226.49..226.49 rows=2 width=4) (actual time=12.219..12.219 rows=1 loops=1)
               Buckets: 131072  Batches: 1  Memory Usage: 1kB
               ->  Hash Join  (cost=69.66..226.49 rows=4 width=4) (actual time=7.681..12.211 rows=1 loops=1)
                     Hash Cond: (cl1.relnamespace = n1.oid)
                     Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
                     ->  Hash Join  (cost=68.56..225.34 rows=4 width=8) (actual time=7.576..11.758 rows=1 loops=1)
                           Hash Cond: (pp.oid = pr.paroid)
                           Join Filter: CASE WHEN (hashed SubPlan 1) THEN (('2024-04-11 13:27:10.386977+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::time
stamp without time zone) AND ('2024-04-11 13:27:10.386977+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)) ELSE false END
                           Rows Removed by Join Filter: 2
                           Extra Text: Hash chain length 2.0 avg, 4 max, using 1002 of 32768 buckets.
                           ->  Hash Join  (cost=7.01..163.72 rows=4 width=40) (actual time=0.543..4.382 rows=2 loops=1)
                                 Hash Cond: (cl1.oid = pp.parrelid)
                                 Extra Text: Hash chain length 1.0 avg, 2 max, using 1008 of 65536 buckets.
                                 ->  Seq Scan on pg_class cl1  (cost=0.00..156.68 rows=1 width=8) (actual time=0.009..3.494 rows=2 loops=1)
                                       Filter: (relname = 'db_files_history'::name)
                                       Rows Removed by Filter: 3514
                                 ->  Hash  (cost=7.00..7.00 rows=1 width=40) (actual time=0.483..0.483 rows=1015 loops=1)
                                       Buckets: 65536  Batches: 1  Memory Usage: 64kB
                                       ->  Seq Scan on pg_partition pp  (cost=0.00..7.00 rows=1 width=40) (actual time=0.027..0.349 rows=1015 loops=1)
                                             Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
                           ->  Hash  (cost=58.00..58.00 rows=1 width=72) (actual time=1.086..1.086 rows=2029 loops=1)
                                 Buckets: 32768  Batches: 1  Memory Usage: 721kB
                                 ->  Seq Scan on pg_partition_rule pr  (cost=0.00..58.00 rows=1 width=72) (actual time=0.007..0.718 rows=2029 loops=1)
                                       Filter: parrangestartincl
                                       Rows Removed by Filter: 1
                           SubPlan 1
                             ->  Materialize  (cost=0.00..3.54 rows=4 width=4) (actual time=0.024..0.025 rows=3 loops=1)
                                   ->  Seq Scan on pg_opclass  (cost=0.00..3.53 rows=4 width=4) (actual time=0.008..0.020 rows=3 loops=1)
                                         Filter: (opcname = 'timestamp_ops'::name)
                                         Rows Removed by Filter: 199
                     ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.005..0.005 rows=1 loops=1)
                           Buckets: 131072  Batches: 1  Memory Usage: 1kB
                           ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4) (actual time=0.003..0.004 rows=1 loops=1)
                                 Filter: (nspname = 'arenadata_toolkit'::name)
                                 Rows Removed by Filter: 7
   ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.016..0.016 rows=1 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1kB
         ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4) (actual time=0.011..0.012 rows=1 loops=1)
               Filter: (nspname = 'arenadata_toolkit'::name)
               Rows Removed by Filter: 7
 Planning time: 4.845 ms
   (slice0)    Executor memory: 16713K bytes.  Work_mem: 721K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 16.642 ms
(52 rows)

postgres=# 
postgres=# explain analyze SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                                JOIN pg_opclass po on pp.parclass[0] = po.oid
                WHERE
                        CASE
                            WHEN
                                                        po.opcname = 'timestamp_ops'
                            THEN now() BETWEEN 
                                        CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                        AND
                                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                            ELSE FALSE
                        END
                                                AND
                                                cl1.relname = 'db_files_history'
                                                AND
                        n1.nspname = 'arenadata_toolkit'
                                                AND
                        n2.nspname = 'arenadata_toolkit'
                        AND
                        pp.paristemplate = false
                        AND
                        pp.parkind = 'r'  --range
                        AND
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                                 QUERY PLA
N                                                                                                                                                                                                                                  
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=227.93..391.92 rows=8 width=0) (actual time=27.624..144.992 rows=1 loops=1)
   Hash Cond: (pp.parrelid = cl1.oid)
   Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
   ->  Hash Join  (cost=70.07..233.94 rows=8 width=4) (actual time=21.849..138.595 rows=1002 loops=1)
         Hash Cond: (pr.paroid = pp.oid)
         Join Filter: CASE WHEN (po.opcname = 'timestamp_ops'::name) THEN (('2024-04-11 13:27:32.137083+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::time
stamp without time zone) AND ('2024-04-11 13:27:32.137083+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)) ELSE false END
         Rows Removed by Join Filter: 1025
         Extra Text: Hash chain length 1.0 avg, 2 max, using 1002 of 32768 buckets.
         ->  Hash Join  (cost=59.11..222.86 rows=4 width=72) (actual time=5.912..10.990 rows=2027 loops=1)
               Hash Cond: (cl2.relnamespace = n2.oid)
               Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
               ->  Hash Join  (cost=58.01..221.68 rows=13 width=76) (actual time=5.778..10.001 rows=2029 loops=1)
                     Hash Cond: (cl2.oid = pr.parchildrelid)
                     Extra Text: Hash chain length 1.0 avg, 3 max, using 1961 of 32768 buckets.
                     ->  Seq Scan on pg_class cl2  (cost=0.00..142.94 rows=5494 width=8) (actual time=0.016..2.098 rows=3516 loops=1)
                     ->  Hash  (cost=58.00..58.00 rows=1 width=72) (actual time=5.609..5.609 rows=2029 loops=1)
                           Buckets: 32768  Batches: 1  Memory Usage: 721kB
                           ->  Seq Scan on pg_partition_rule pr  (cost=0.00..58.00 rows=1 width=72) (actual time=0.014..1.777 rows=2029 loops=1)
                                 Filter: parrangestartincl
                                 Rows Removed by Filter: 1
               ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.022..0.022 rows=1 loops=1)
                     Buckets: 131072  Batches: 1  Memory Usage: 1kB
                     ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4) (actual time=0.019..0.019 rows=1 loops=1)
                           Filter: (nspname = 'arenadata_toolkit'::name)
                           Rows Removed by Filter: 7
         ->  Hash  (cost=10.87..10.87 rows=3 width=72) (actual time=1.459..1.459 rows=1015 loops=1)
               Buckets: 32768  Batches: 1  Memory Usage: 96kB
               ->  Hash Join  (cost=7.01..10.87 rows=8 width=72) (actual time=0.706..1.161 rows=1015 loops=1)
                     Hash Cond: (po.oid = pp.parclass[0])
                     Extra Text: Hash chain length 507.5 avg, 1002 max, using 2 of 65536 buckets.
                     ->  Seq Scan on pg_opclass po  (cost=0.00..3.02 rows=202 width=68) (actual time=0.015..0.072 rows=202 loops=1)
                     ->  Hash  (cost=7.00..7.00 rows=1 width=40) (actual time=0.624..0.624 rows=1015 loops=1)
                           Buckets: 65536  Batches: 1  Memory Usage: 64kB
                           ->  Seq Scan on pg_partition pp  (cost=0.00..7.00 rows=1 width=40) (actual time=0.017..0.368 rows=1015 loops=1)
                                 Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
   ->  Hash  (cost=157.81..157.81 rows=2 width=4) (actual time=5.570..5.570 rows=1 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1kB
         ->  Hash Join  (cost=1.10..157.81 rows=4 width=4) (actual time=0.535..5.541 rows=1 loops=1)
               Hash Cond: (cl1.relnamespace = n1.oid)
               Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
               ->  Seq Scan on pg_class cl1  (cost=0.00..156.68 rows=1 width=8) (actual time=0.103..4.324 rows=2 loops=1)
                     Filter: (relname = 'db_files_history'::name)
                     Rows Removed by Filter: 3514
               ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.076..0.076 rows=1 loops=1)
                     Buckets: 131072  Batches: 1  Memory Usage: 1kB
                     ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4) (actual time=0.055..0.057 rows=1 loops=1)
                           Filter: (nspname = 'arenadata_toolkit'::name)
                           Rows Removed by Filter: 7
 Planning time: 9.686 ms
   (slice0)    Executor memory: 14449K bytes.  Work_mem: 721K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 145.421 ms
(53 rows)

postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# explain analyze SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                WHERE                                                        
                        CASE
                            WHEN
                                                        pp.parclass[0] IN (SELECT oid FROM pg_opclass WHERE opcname = 'timestamp_ops')
                            THEN now() BETWEEN 
                                        CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                        AND                                    
                                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                            ELSE FALSE                                         
                        END           
                                                AND
                                                cl1.relname = 'db_files_history'
                                                AND                             
                        n1.nspname = 'arenadata_toolkit'
                                                AND
                        n2.nspname = 'arenadata_toolkit'
                        AND                             
                        pp.paristemplate = false
                        AND                     
                        pp.parkind = 'r'  --range
                        AND                      
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                                 QUERY PLA
N                                                                                                                                                                                                                                  
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=227.63..391.26 rows=4 width=0) (actual time=6.807..8.849 rows=1 loops=1)
   Hash Cond: (cl2.relnamespace = n2.oid)
   Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
   ->  Hash Join  (cost=226.53..390.11 rows=4 width=4) (actual time=6.715..8.514 rows=1 loops=1)
         Hash Cond: (cl2.oid = pr.parchildrelid)
         Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
         ->  Seq Scan on pg_class cl2  (cost=0.00..142.94 rows=5494 width=8) (actual time=0.008..0.870 rows=3516 loops=1)
         ->  Hash  (cost=226.49..226.49 rows=2 width=4) (actual time=6.604..6.604 rows=1 loops=1)
               Buckets: 131072  Batches: 1  Memory Usage: 1kB
               ->  Hash Join  (cost=69.66..226.49 rows=4 width=4) (actual time=5.639..6.602 rows=1 loops=1)
                     Hash Cond: (cl1.relnamespace = n1.oid)
                     Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
                     ->  Hash Join  (cost=68.56..225.34 rows=4 width=8) (actual time=5.522..6.271 rows=1 loops=1)
                           Hash Cond: (pp.oid = pr.paroid)
                           Join Filter: CASE WHEN (hashed SubPlan 1) THEN (('2024-04-11 13:37:46.439345+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::time
stamp without time zone) AND ('2024-04-11 13:37:46.439345+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)) ELSE false END
                           Rows Removed by Join Filter: 2
                           Extra Text: Hash chain length 2.0 avg, 4 max, using 1002 of 32768 buckets.
                           ->  Hash Join  (cost=7.01..163.72 rows=4 width=40) (actual time=0.556..1.140 rows=2 loops=1)
                                 Hash Cond: (cl1.oid = pp.parrelid)
                                 Extra Text: Hash chain length 1.0 avg, 2 max, using 1008 of 65536 buckets.
                                 ->  Seq Scan on pg_class cl1  (cost=0.00..156.68 rows=1 width=8) (actual time=0.011..0.484 rows=2 loops=1)
                                       Filter: (relname = 'db_files_history'::name)
                                       Rows Removed by Filter: 3514
                                 ->  Hash  (cost=7.00..7.00 rows=1 width=40) (actual time=0.393..0.393 rows=1015 loops=1)
                                       Buckets: 65536  Batches: 1  Memory Usage: 64kB
                                       ->  Seq Scan on pg_partition pp  (cost=0.00..7.00 rows=1 width=40) (actual time=0.006..0.253 rows=1015 loops=1)
                                             Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
                           ->  Hash  (cost=58.00..58.00 rows=1 width=72) (actual time=0.946..0.946 rows=2029 loops=1)
                                 Buckets: 32768  Batches: 1  Memory Usage: 721kB
                                 ->  Seq Scan on pg_partition_rule pr  (cost=0.00..58.00 rows=1 width=72) (actual time=0.005..0.538 rows=2029 loops=1)
                                       Filter: parrangestartincl
                                       Rows Removed by Filter: 1
                           SubPlan 1
                             ->  Materialize  (cost=0.00..3.54 rows=4 width=4) (actual time=0.026..0.028 rows=3 loops=1)
                                   ->  Seq Scan on pg_opclass  (cost=0.00..3.53 rows=4 width=4) (actual time=0.009..0.023 rows=3 loops=1)
                                         Filter: (opcname = 'timestamp_ops'::name)
                                         Rows Removed by Filter: 199
                     ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.007..0.007 rows=1 loops=1)
                           Buckets: 131072  Batches: 1  Memory Usage: 1kB
                           ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4) (actual time=0.005..0.005 rows=1 loops=1)
                                 Filter: (nspname = 'arenadata_toolkit'::name)
                                 Rows Removed by Filter: 7
   ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.013..0.013 rows=1 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1kB
         ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4) (actual time=0.009..0.010 rows=1 loops=1)
               Filter: (nspname = 'arenadata_toolkit'::name)
               Rows Removed by Filter: 7
 Planning time: 3.925 ms
   (slice0)    Executor memory: 16713K bytes.  Work_mem: 721K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 9.024 ms
(52 rows)

postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# 
postgres=# explain analyze SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                                JOIN pg_opclass po on pp.parclass[0] = po.oid
                WHERE
                        CASE
                            WHEN                                                                                                      
                                                        po.opcname = 'timestamp_ops'
                            THEN now() BETWEEN                                                                                 
                                        CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                        AND                                                                                  
                                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                            ELSE FALSE
                        END                        
                                                AND                             
                                                cl1.relname = 'db_files_history'
                                                AND
                        n1.nspname = 'arenadata_toolkit'
                                                AND
                        n2.nspname = 'arenadata_toolkit'
                        AND                     
                        pp.paristemplate = false
                        AND                      
                        pp.parkind = 'r'  --range
                        AND                        
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                                 QUERY PLA
N                                                                                                                                                                                                                                  
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=227.93..391.92 rows=8 width=0) (actual time=7.115..118.567 rows=1 loops=1)
   Hash Cond: (pp.parrelid = cl1.oid)
   Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
   ->  Hash Join  (cost=70.07..233.94 rows=8 width=4) (actual time=6.165..117.082 rows=1002 loops=1)
         Hash Cond: (pr.paroid = pp.oid)
         Join Filter: CASE WHEN (po.opcname = 'timestamp_ops'::name) THEN (('2024-04-11 13:37:58.984063+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::time
stamp without time zone) AND ('2024-04-11 13:37:58.984063+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)) ELSE false END
         Rows Removed by Join Filter: 1025
         Extra Text: Hash chain length 1.0 avg, 2 max, using 1002 of 32768 buckets.
         ->  Hash Join  (cost=59.11..222.86 rows=4 width=72) (actual time=1.142..4.976 rows=2027 loops=1)
               Hash Cond: (cl2.relnamespace = n2.oid)
               Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
               ->  Hash Join  (cost=58.01..221.68 rows=13 width=76) (actual time=1.050..4.016 rows=2029 loops=1)
                     Hash Cond: (cl2.oid = pr.parchildrelid)
                     Extra Text: Hash chain length 1.0 avg, 3 max, using 1961 of 32768 buckets.
                     ->  Seq Scan on pg_class cl2  (cost=0.00..142.94 rows=5494 width=8) (actual time=0.009..1.265 rows=3516 loops=1)
                     ->  Hash  (cost=58.00..58.00 rows=1 width=72) (actual time=1.001..1.001 rows=2029 loops=1)
                           Buckets: 32768  Batches: 1  Memory Usage: 721kB
                           ->  Seq Scan on pg_partition_rule pr  (cost=0.00..58.00 rows=1 width=72) (actual time=0.005..0.555 rows=2029 loops=1)
                                 Filter: parrangestartincl
                                 Rows Removed by Filter: 1
               ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.007..0.007 rows=1 loops=1)
                     Buckets: 131072  Batches: 1  Memory Usage: 1kB
                     ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4) (actual time=0.006..0.006 rows=1 loops=1)
                           Filter: (nspname = 'arenadata_toolkit'::name)
                           Rows Removed by Filter: 7
         ->  Hash  (cost=10.87..10.87 rows=3 width=72) (actual time=0.989..0.989 rows=1015 loops=1)
               Buckets: 32768  Batches: 1  Memory Usage: 96kB
               ->  Hash Join  (cost=7.01..10.87 rows=8 width=72) (actual time=0.479..0.821 rows=1015 loops=1)
                     Hash Cond: (po.oid = pp.parclass[0])
                     Extra Text: Hash chain length 507.5 avg, 1002 max, using 2 of 65536 buckets.
                     ->  Seq Scan on pg_opclass po  (cost=0.00..3.02 rows=202 width=68) (actual time=0.005..0.031 rows=202 loops=1)
                     ->  Hash  (cost=7.00..7.00 rows=1 width=40) (actual time=0.426..0.426 rows=1015 loops=1)
                           Buckets: 65536  Batches: 1  Memory Usage: 64kB
                           ->  Seq Scan on pg_partition pp  (cost=0.00..7.00 rows=1 width=40) (actual time=0.005..0.239 rows=1015 loops=1)
                                 Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
   ->  Hash  (cost=157.81..157.81 rows=2 width=4) (actual time=0.831..0.831 rows=1 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1kB
         ->  Hash Join  (cost=1.10..157.81 rows=4 width=4) (actual time=0.109..0.830 rows=1 loops=1)
               Hash Cond: (cl1.relnamespace = n1.oid)
               Extra Text: Hash chain length 1.0 avg, 1 max, using 1 of 131072 buckets.
               ->  Seq Scan on pg_class cl1  (cost=0.00..156.68 rows=1 width=8) (actual time=0.018..0.496 rows=2 loops=1)
                     Filter: (relname = 'db_files_history'::name)
                     Rows Removed by Filter: 3514
               ->  Hash  (cost=1.09..1.09 rows=1 width=4) (actual time=0.007..0.007 rows=1 loops=1)
                     Buckets: 131072  Batches: 1  Memory Usage: 1kB
                     ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4) (actual time=0.004..0.004 rows=1 loops=1)
                           Filter: (nspname = 'arenadata_toolkit'::name)
                           Rows Removed by Filter: 7
 Planning time: 5.185 ms
   (slice0)    Executor memory: 14449K bytes.  Work_mem: 721K bytes max.
 Memory used:  128000kB
 Optimizer: Postgres query optimizer
 Execution time: 118.728 ms
(53 rows)

postgres=# 
 

```

</details>

Do you need any other experiment?